### ISSUE TITLE
run  ./build.sh wrong

### ISSUE THREAD

**Issue by Clarence-wcz** [microsoft/onnxruntime#1474]
**Describe the bug**

**Urgency**
If there are particular important use cases blocked by this or strict project-related timelines, please share more information and dates. If there are no hard deadlines, please specify none.

**System information**
- OS Platform and Distribution:Ubuntu 14.04
- ONNX Runtime installed from (source or binary):
- ONNX Runtime version:
- Python version:3.6.1
- Visual Studio version (if applicable):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

Traceback (most recent call last):
  File "/home/linux/onnxruntime-master/tools/ci_build/build.py", line 902, in <module>
    sys.exit(main())
  File "/home/linux/onnxruntime-master/tools/ci_build/build.py", line 845, in main
    args, cmake_extra_args)
  File "/home/linux/onnxruntime-master/tools/ci_build/build.py", line 390, in generate_build_tree
    run_subprocess(cmake_args  + ["-DCMAKE_BUILD_TYPE={}".format(config)], cwd=config_build_dir)
  File "/home/linux/onnxruntime-master/tools/ci_build/build.py", line 181, in run_subprocess
    return subprocess.run(args, cwd=cwd, check=True, stdout=stdout, stderr=stderr, env=my_env, shell=shell)
  File "/usr/local/lib/python3.6/subprocess.py", line 418, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['/usr/local/bin/cmake', '/home/linux/onnxruntime-master/cmake', '-Donnxruntime_RUN_ONNX_TESTS=OFF', '-Donnxruntime_GENERATE_TEST_REPORTS=ON', '-Donnxruntime_DEV_MODE=ON', '-DPYTHON_EXECUTABLE=/usr/local/bin/python3', '-Donnxruntime_USE_CUDA=OFF', '-Donnxruntime_USE_NSYNC=OFF', '-Donnxruntime_CUDNN_HOME=', '-Donnxruntime_CUDA_HOME=', '-Donnxruntime_USE_JEMALLOC=OFF', '-Donnxruntime_ENABLE_PYTHON=ON', '-Donnxruntime_BUILD_CSHARP=OFF', '-Donnxruntime_BUILD_SHARED_LIB=OFF', '-Donnxruntime_USE_EIGEN_FOR_BLAS=ON', '-Donnxruntime_USE_OPENBLAS=OFF', '-Donnxruntime_USE_MKLDNN=OFF', '-Donnxruntime_USE_MKLML=OFF', '-Donnxruntime_USE_NGRAPH=OFF', '-Donnxruntime_USE_OPENVINO=OFF', '-Donnxruntime_USE_OPENVINO_BINARY=OFF', '-Donnxruntime_USE_OPENVINO_SOURCE=OFF', '-Donnxruntime_USE_OPENVINO_MYRIAD=OFF', '-Donnxruntime_USE_OPENVINO_GPU_FP32=OFF', '-Donnxruntime_USE_OPENVINO_GPU_FP16=OFF', '-Donnxruntime_USE_OPENVINO_CPU_FP32=OFF', '-Donnxruntime_USE_OPENVINO_VAD_R=OFF', '-Donnxruntime_USE_OPENMP=ON', '-Donnxruntime_USE_TVM=OFF', '-Donnxruntime_USE_LLVM=OFF', '-Donnxruntime_ENABLE_MICROSOFT_INTERNAL=OFF', '-Donnxruntime_USE_BRAINSLICE=OFF', '-Donnxruntime_USE_NUPHAR=OFF', '-Donnxruntime_USE_EIGEN_THREADPOOL=OFF', '-Donnxruntime_USE_TENSORRT=OFF', '-Donnxruntime_TENSORRT_HOME=', '-Donnxruntime_CROSS_COMPILING=OFF', '-Donnxruntime_BUILD_SERVER=OFF', '-Donnxruntime_BUILD_x86=ON', '-Donnxruntime_USE_FULL_PROTOBUF=OFF', '-Donnxruntime_DISABLE_CONTRIB_OPS=OFF', '-Donnxruntime_MSVC_STATIC_RUNTIME=OFF', '-Donnxruntime_ENABLE_LANGUAGE_INTEROP_OPS=OFF', '-DCMAKE_BUILD_TYPE=RelWithDebInfo']' returned non-zero exit status 1.

**Comment by Clarence-wcz** [comment]
How can I solve this issue?

**Comment by snnn** [comment]
Could you please post the full log here? You can upload the file as an attachment.

**Comment by Clarence-wcz** [comment]
[CMakeError.log](https://github.com/microsoft/onnxruntime/files/3424665/CMakeError.log)

**Comment by Clarence-wcz** [comment]
> Could you please post the full log here? You can upload the file as an attachment.

Thanks! I have uploaded the CmakeError.log

**Comment by snnn** [comment]
```bash
./build.sh >& 1.log
```
Then post 1.log here.

**Comment by Clarence-wcz** [comment]
> ```shell
> ./build.sh >& 1.log
> ```
> 
> Then post 1.log here.

[1.log](https://github.com/microsoft/onnxruntime/files/3425677/1.log)
TThanks.

**Comment by Clarence-wcz** [comment]
> ```shell
> ./build.sh >& 1.log
> ```
> 
> Then post 1.log here.

[1.log](https://github.com/microsoft/onnxruntime/files/3425679/1.log)
Thanks.

**Comment by Clarence-wcz** [comment]
> Please manually run
> 
> ```shell
> git submodule update --init --recursive
> ```
> 
> Before running build.sh. And please check if the folder "/home/linux/onnxruntime/cmake/external/protobuf" exists before running build.sh.
> 
> Seems like the command failed.
I have solved this problem, but test is failed. I can't get the wheel under dist folder.
[2.log](https://github.com/microsoft/onnxruntime/files/3429377/2.log)

**Comment by Clarence-wcz** [comment]
> Please manually run
> 
> ```shell
> git submodule update --init --recursive
> ```
> 
> Before running build.sh. And please check if the folder "/home/linux/onnxruntime/cmake/external/protobuf" exists before running build.sh.
> 
> Seems like the command failed.

Thanks for your help! I have solved all problems.

**Comment by snnn** [comment]
The log says:

 [==========] 1100 tests from 122 test cases ran. (6735 ms total)
 [  PASSED  ] 1098 tests.
 [  FAILED  ] 2 tests, listed below:
 [  FAILED  ] RoiAlignTest.AvgModeNegativeInvalidNumRoiDims
 [  FAILED  ] RoiAlignTest.MismatchNumRois

**Comment by jywu-msft** [comment]
> > Please manually run
> > ```shell
> > git submodule update --init --recursive
> > ```
> > 
> > 
> > Before running build.sh. And please check if the folder "/home/linux/onnxruntime/cmake/external/protobuf" exists before running build.sh.
> > Seems like the command failed.
> > I have solved this problem, but test is failed. I can't get the wheel under dist folder.
> > [2.log](https://github.com/microsoft/onnxruntime/files/3429377/2.log)

skip the tests (as a workaround to get the wheel built)
by adding --update --build to your ./build.sh arguments

**Comment by Clarence-wcz** [comment]
> The log says:
> 
> [==========] 1100 tests from 122 test cases ran. (6735 ms total)
> [ PASSED ] 1098 tests.
> [ FAILED ] 2 tests, listed below:
> [ FAILED ] RoiAlignTest.AvgModeNegativeInvalidNumRoiDims
> [ FAILED ] RoiAlignTest.MismatchNumRois

Thanks! But now I have a new problem. When I used the "CXX_Api_Sample.cpp"  to test, there were a lot of errors with undefined references , I had added the compiled static library file. For example, "/home/linux/onnxruntime/build/Linux/Debug/libonnxruntime_providers.a(reverse_sequence.cc.o): In function `void onnxruntime::ReverseSequenceImpl<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(onnxruntime::Tensor const&, onnxruntime::Tensor&, gsl::span<long const, -1l>, long, long, long, bool) [clone ._omp_fn.27]':
/home/linux/onnxruntime/onnxruntime/core/providers/cpu/tensor/reverse_sequence.cc:141: undefined reference to `omp_get_num_threads'
/home/linux/onnxruntime/onnxruntime/core/providers/cpu/tensor/reverse_sequence.cc:141: undefined reference to `omp_get_thread_num' "
Can you help me?

**Comment by Clarence-wcz** [comment]
> > > Please manually run
> > > ```shell
> > > git submodule update --init --recursive
> > > ```
> > > 
> > > 
> > > Before running build.sh. And please check if the folder "/home/linux/onnxruntime/cmake/external/protobuf" exists before running build.sh.
> > > Seems like the command failed.
> > > I have solved this problem, but test is failed. I can't get the wheel under dist folder.
> > > [2.log](https://github.com/microsoft/onnxruntime/files/3429377/2.log)
> 
> skip the tests (as a workaround to get the wheel built)
> by adding --update --build to your ./build.sh arguments

Thanks! I have solved this problem.

**Comment by snnn** [comment]
> there were a lot of errors with undefined references

You need to link the code to libonnxruntime.so, not the static libs. 
Run `./build.sh --config RelWithDebInfo --build_shared_lib --parallel` to get it.

**Comment by Clarence-wcz** [comment]
> > there were a lot of errors with undefined references
> 
> You need to link the code to libonnxruntime.so, not the static libs.
> Run `./build.sh --config RelWithDebInfo --build_shared_lib --parallel` to get it.

Thanks ！I had solve this problem. But I used the command "./build.sh --config Debug --use_preinstalled_eigen --eigen_path /home/linux/eigen_lib/include/eigen3/ --use_eigenthreadpool  --build_shared_lib --update --build" to get eigen-version , it is failed. And other question is eigen library has changed in the latest version, you might need to update. For example, #include "Eigen/src/Core/arch/GPU/Half.h" , it needs to be modified to "xx/CUDA/Half.h" in cast_op,cc.
[4.log](https://github.com/microsoft/onnxruntime/files/3442285/4.log)

**Comment by snnn** [comment]
Hi,

Could you please remove the ' --use_preinstalled_eigen' flag?  onnxruntime only works with one specific eigen version, unfortunately it's not the one you're having.

**Comment by Clarence-wcz** [comment]
Could you tell me the specific version number of Eigen? I may need to tweak the Eigen version used by my project. Thanks!

**Comment by Clarence-wcz** [comment]
> Hi,
> 
> Could you please remove the ' --use_preinstalled_eigen' flag? onnxruntime only works with one specific eigen version, unfortunately it's not the one you're having.

Could you tell me the specific version number of Eigen? I may need to tweak the Eigen version used by my project. Thanks!

**Comment by snnn** [comment]
Hi @Clarence-wcz, 

<hr/>
C:\src\onnxruntime>cd cmake\external\eigen

C:\src\onnxruntime\cmake\external\eigen>git log
commit 1456fe2fac0cd074bbbb8630ba003a7c325d015b (HEAD)
Author: Eugene Zhulenev <ezhulenev@google.com>
Date:   Tue Apr 2 15:04:26 2019 -0700

Add missing semicolon
<hr/>

BTW, you are free to use any eigen version in your source code, it won't conflict with onnxruntime's eigen. The only thing need be caution is: if you want to build onnxruntime from source code, it's better to do it in a clean docker image, and that image should only have one eigen in the whole filesystem.

**Comment by snnn** [comment]
Close this issue because the original problem is solved.  If you have more questions, please feel free to create a new issue.

**Comment by snnn** [comment]
Removed the bug  label. Not something wrong in onnxruntime.
